### PR TITLE
Add ComputeDerivativeGroup*NV capabilities to trim capabilities pass.

### PR DESCRIPTION
**Pull request by s-perron**

### CODE REVIEW

**Review by Keenuts — COMMENTED** [review]
I think this is incomplete:
```
 If neither derivative group mode was specified, the derivatives return zero.
```

Seems like removing the execution mode, but having the capability does have an effect: it allows the sampling instructions in `GLCompute`.
So if such instruction is used in a compute shader, but the capability is removed, we'd produce the wrong module no?

File: source/opt/aggressive_dead_code_elim_pass.cpp [hunk 1]
```diff
@@ -941,67 +941,40 @@ Pass::Status AggressiveDCEPass::Process() {
...
-      "SPV_KHR_integer_dot_product",
-      "SPV_EXT_shader_image_int64",
-      "SPV_KHR_non_semantic_info",
-      "SPV_KHR_uniform_group_instructions",
-      "SPV_KHR_fragment_shader_barycentric",
-      "SPV_NV_bindless_texture",
-      "SPV_EXT_shader_atomic_float_add",
-      "SPV_EXT_fragment_shader_interlock",
-  });
+  extensions_allowlist_.insert(
```
[the remark this answers is not in the document]
Comment (s-perron) [reply source/opt/aggressive_dead_code_elim_pass.cpp:944]: Yes, that is clang format.

**Comment by s-perron** [comment]
> I think this is incomplete:
> 
> ```
>  If neither derivative group mode was specified, the derivatives return zero.
> ```
> 
> Seems like removing the execution mode, but having the capability does have an effect: it allows the sampling instructions in `GLCompute`. So if such instruction is used in a compute shader, but the capability is removed, we'd produce the wrong module no?

I missed that statement in the spec. When I tried cases without the execution mode the validator complained:

```
~/test » spirv-val t.spv                                                                                                                                                                                                                               stevenperron@stevenperron-cloudtop
error: line 11: OpEntryPoint Entry Point <id> '1[%1]'s callgraph contains function <id> '1[%1]', which cannot be used with the current execution modes:
Derivative instructions require DerivativeGroupQuadsNV or DerivativeGroupLinearNV execution mode for GLCompute execution model: DPdx

  %1 = OpFunction %void None %5

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/test » spirv-dis t.spv                                                                                                                                                                                                                           1 ↵ stevenperron@stevenperron-cloudtop
; SPIR-V
; Version: 1.3
; Generator: Khronos SPIR-V Tools Assembler; 0
; Bound: 8
; Schema: 0
               OpCapability ComputeDerivativeGroupQuadsNV
               OpCapability ComputeDerivativeGroupLinearNV
               OpCapability Shader
               OpExtension "SPV_NV_compute_shader_derivatives"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %1 "main"
       %void = OpTypeVoid
      %float = OpTypeFloat 32
    %float_0 = OpConstant %float 0
          %5 = OpTypeFunction %void
          %1 = OpFunction %void None %5
          %6 = OpLabel
          %7 = OpDPdx %float %float_0
               OpReturn
               OpFunctionEnd

```

**Comment by s-perron** [comment]
In fact, that has been the behaviour since the validation for the extension was first implemented by Nvidia: https://github.com/KhronosGroup/SPIRV-Tools/commit/bac82f49aa106f7b440e882d593db31faa7d538a#diff-1232a89db063a58c947bb0e253a3161c55e2c5e49666aa96d55aaa6d59b0f8fdR78-R90

**Comment by s-perron** [comment]
@dgkoch @jmacnak-nv Do either of you have any insight into the difference between the spec and the validator?

**Comment by s-perron** [comment]
@Keenuts I suggest we merge this, since this behaviour matches the validator, and we have not heard anything from NV. I have opened up a follow up issue.

**Comment by Keenuts** [comment]
As long as the validator refuses to accept cases where we might remove the extension when required, that's fine for me.

**Review by Keenuts — CHANGES_REQUESTED** [review]

**Review by Keenuts — APPROVED** [review]
LGTM for the change.
Now, it's up to you how to handle the validator issue.
Since both GLSL and the SPIR-V spec are consistent, maybe fixing the validator to correctly handle this case could be useful, but in the if this is required to unblock something, I'm fine with this temporary solution until it's rewritten with the validation rule.

**Pull request merged**